### PR TITLE
[Triton-MLIR][HSACO] Added HSACO generation.

### PR DESCRIPTION
**Pull request by dfukalov**
The change is based on Rohit' commit 9bf807f310b19f6596526f84baf8bde46ff2dbd2

Used `AMDGCN_HSACO_DUMP_PATH=<dumpfile>` (example usage can be found in `tritongpu_to_hsaco.mlir` test) notion to save the copy of temp generated HSACO file. So a test is able to pass it to `llvm-readobj` and check its output.

### CODE REVIEW

File: lib/Target/CMakeLists.txt
Comment (rsanthanam-amd) [inline lib/Target/CMakeLists.txt:1]: Since this PR already takes care of AMDGCN generation, please remove this subdirectory and references to it.

This was added in a previous commit.

File: lib/Target/LLVMIR/LLVMIRTranslation.cpp
Comment (whchung) [inline lib/Target/LLVMIR/LLVMIRTranslation.cpp:25]: this header file is not really needed?

File: lib/Target/HSACO/HSACOTranslation.cpp
Comment (whchung) [inline lib/Target/HSACO/HSACOTranslation.cpp:44]: Do we need `LLVMInitializeAMDGPUAsmParser()` here?

**Comment by whchung** [comment]
Now #34 was just merged, perhaps this PR deserves a rebase to ensure no conflicts or mismatched functionalities.

File: lib/Target/HSACO/HSACOTranslation.cpp
Comment (whchung) [inline lib/Target/HSACO/HSACOTranslation.cpp:36]: do we really need `regex`, `cstdio`, `iostream`? from the look of the PR these headers aren't really used.

File: lib/Target/HSACO/HSACOTranslation.cpp
Comment (whchung) [inline lib/Target/HSACO/HSACOTranslation.cpp:99]: please use `raw_ostream` instead of `std::cout`.

File: lib/Target/HSACO/HSACOTranslation.cpp
Comment (whchung) [inline lib/Target/HSACO/HSACOTranslation.cpp:110]: please use `raw_ostream` instead of `std::cout`.

File: lib/Target/HSACO/HSACOTranslation.cpp
Comment (whchung) [inline lib/Target/HSACO/HSACOTranslation.cpp:126]: please use `raw_ostream` instead of `std::cout`.

**Review by whchung — CHANGES_REQUESTED** [review]
Please reuse `raw_ostream` instances already created by LLVM such as `llvm::outs` and `llvm::errs` for logging.

**Comment by dfukalov** [comment]
> Now https://github.com/ROCmSoftwarePlatform/triton/pull/34 was just merged, perhaps this PR deserves a rebase to ensure no conflicts or mismatched functionalities.

@whchung the PR was actually merged in different branch. I'm agree that it conflicts with the changes, but suggest to resolve it with separate future step(s).

**Comment by dfukalov** [comment]
> Since this PR already takes care of AMDGCN generation, please remove this subdirectory and references to it.
> This was added in a previous commit.
@rsanthanam-amd Thanks for this point, I missed it. But It seems we should actually use "AMDGCN" or "AMDGPU" name for the `Target` subfolder (since there is "PTX" but not "CUBIN". If you're agree, I'll merge new parts into "AMDGCN" subfolder.

**Comment by rsanthanam-amd** [comment]
> > Since this PR already takes care of AMDGCN generation, please remove this subdirectory and references to it.
> > This was added in a previous commit.
> > @rsanthanam-amd Thanks for this point, I missed it. But It seems we should actually use "AMDGCN" or "AMDGPU" name for the `Target` subfolder (since there is "PTX" but not "CUBIN". If you're agree, I'll merge new parts into "AMDGCN" subfolder.

Yeah, i'm ok with your suggestion.

**Review by rsanthanam-amd — APPROVED** [review]

**Review by rsanthanam-amd — CHANGES_REQUESTED** [review]
Please address comments.

**Review by rsanthanam-amd — APPROVED** [review]
My comments have been addressed.

Waiting for Jack to approve before merging.

**Review by whchung — APPROVED** [review]

**Pull request merged**
